Update MixUp to allow use of index labels (#825)
mosaicml/composer: composer/algorithms/mixup/README.md

|:--:
|*Two different training examples (a picture of a bird and a picture of a frog) that have been combined by MixUp into a single example. The corresponding targets are a convex combination of the targets for the bird class and the frog class.*|

## How to Use

### Functional Interface

> Running `mixup` using index labels and interpolating the loss (a trick when using cross entropy)
> ```python
> # Run the mixup algorithm directly on the inputs and targets using the Composer functional API
> 

Reviewer: ```suggestion
Running `mixup` using index labels and interpolating the loss (a trick when using cross entropy):
```python
```

Author: Fixed.

import composer.functional as cf

def training_loop(model, train_loader):
  opt = torch.optim.Adam(model.parameters())
  loss_fn = F.cross_entropy
  model.train()

  for epoch in range(num_epochs):
      for X, y in train_loader:
          X_mixed, y_orig, y_perm, mixing = cf.mixup_batch(X, y, alpha=0.2)
          y_hat = model(X_mixed)
          loss = (1 - mixing) * loss_fn(y_hat, y_orig) + mixing * loss_fn(y_hat, y_perm)
          loss.backward()
          opt.step()
          opt.zero_grad()
```

> Running `mixup` using dense/one-hot labels and interpolating the labels (general case)
> ```python
> # Run the mixup algorithm directly on the inputs and targets using the Composer functional API
> 

Reviewer: ```suggestion
Running `mixup` using dense/one-hot labels and interpolating the labels (general case):
```python
```

Author: Fixed.

from composer.trainer import Trainer

mixup = MixUp(alpha=0.2, interpolate_loss=True)

trainer = Trainer(model=model,
                  train_dataloader=train_dataloader,
                  max_duration='1ep',
                  algorithms=[mixup])

trainer.fit()
```

Running `mixup` using dense/one-hot labels and interpolating the labels (general case).
```python
from composer.algorithms import MixUp
from composer.trainer import Trainer

mixup = MixUp(alpha=0.2, interpolate_loss=False)

trainer = Trainer(model=model,
                  train_dataloader=train_dataloader,
                  max_duration='1ep',
                  algorithms=[mixup])

trainer.fit()
```

### Implementation Details

Composer's MixUp has two modes of use. When using `interpolate_loss=True` MixUp does not directly interpolate the targets, but rather interpolates the loss function by calling it on the original target and the target of the mixed in examples and interpolating the two loss values. For loss functions that are linear in the targets (such as cross entropy) this is mathematically equivalent to interpolating the targets. Many other implementations of mixup work this way.

> For loss functions that are not linear in the targets, this trick may still be used but results in behavior that differs from the description in the [original paper](https://arxiv.org/abs/1710.09412). If the loss function in use can accept non-index targets (such as dense or one-hot labels) then Composer's MixUp can be used with `interpolate_loss=False` which interpolates the targets of the two samples. This gives behavior matching the usual description of MixUp.

Reviewer: just a comma
```suggestion
For loss functions that are not linear in the targets, this trick may still be used but results in behavior that differs from the description in the [original paper](https://arxiv.org/abs/1710.09412). If the loss function in use can accept non-index targets (such as dense or one-hot labels) then Composer's MixUp can be used with `interpolate_loss=False`, which interpolates the targets of the two samples. This gives behavior matching the usual description of MixUp.
```

> MixUp requires a small amount of additional GPU compute and memory to produce the mixed-up batch.
> In our experiments, we have found these additional resource requirements to be negligible.

> ❗ When `interpolate_loss=False` MixUp Produces a Full Distribution, Not a Target Index
>
> Many classification tasks represent the target value using the index of the target value rather than the full one-hot encoding of the label value.

> > With `interpolate_loss=False` MixUp interpolates between two target values for each example, it must represent the final targets as a dense distribution.

Reviewer: Also distribution -> vector of probabilities in the sentence right after this, and possibly elsewhere in the method card. Or define what we mean by "dense distribution" somewhere. I can ready this as referring to, e.g., a Gaussian, a Uniform, or basically any other continuous distribution, which makes this text seem confusing.

```suggestion
> With `interpolate_loss=False`, MixUp interpolates between two target values for each example, so it must represent the final targets as a dense vector of probabilities.
```

Author: Cleaned up the language here according to your suggestions.

> Our implementation of MixUp turns each label into a dense distribution (if it has not already been converted into a distribution).
> The loss function used for the model must be able to accept this dense distribution as the target.

> >❗ When `interpolate_loss=True` MixUp interpolates the loss rather than the targets.
> 
> > This is fine for loss functions that are linear in the targets, such as cross entropy, but may produce unexpected results for other loss functions.

Reviewer: Currently shows up as two different blocks [in the docs](https://mosaicml-composer--825.com.readthedocs.build/en/825/method_cards/mixup.html).
```suggestion
>❗ When `interpolate_loss=True` MixUp interpolates the loss rather than the targets.
> 
> This is fine for loss functions that are linear in the targets, such as cross entropy, but may produce unexpected results for other loss functions.
```

Reviewer: also, at the bottom of the method card, should "mixup" be capitalized in the paper name?

Author: Fixed the the blocking here. 

For the name, the original paper used "mixup" in all lower case, so for the citation I left it as it appears in the paper. Is that what you meant?

Reviewer: yes; sgtm
